src/box/lua/config/configdata.lua

```diff
@@ -848,6 +848,18 @@ local function validate_multi_master(iconfig_def, peers)
     end
 end
 
+local function validate_group_names_are_unique(cconfig)
+    local groups = {}
+
+    for group_name, _ in pairs(cconfig.groups) do
+        if groups[group_name] then
+            error(('found groups with the same name %q.'):format(group_name), 0)
+        end
+
+        groups[group_name] = true
+    end
+end
```

Comment:
It looks like it is impossible to reach the error: all the keys in a Lua table are unique. We can't receive the same string twice as a key from `pairs()` if the table is not changing during the traversal.

Reply:
I understood why the test cases pass: `instance_names_from_config` returns an empty table and no checks are performed.

Reply:
Thank you for noticing.

I've fixed the tests. See https://github.com/tarantool/tarantool/pull/10913#discussion_r1878174507.

Also, you're right it's unreachable. Turns out currently it's not so simple to detect duplicates in the same subgroup. I've created a separate issue #10917 on that prolem and left a comment that it's not checked in the functions.

```diff
+                -- Currently, it's not possible to handle
+                -- groups, replicasets, instances with the same
+                -- names in the same subgroup. E.g. such cluster
+                -- config are considered ok (gh-10917):
+                -- * g-001
+                --   * r-001
+                --     * i-001
+                --     * i-001 (duplicate)
+
+                -- Duplicating instance name is found within
+                -- distinct replicasets.
```

Now, the patchset only fixes replicasets/instances with the same name behavior.